File: aat-android/src/main/java/ch/bailu/aat/views/preferences/SolidImageButton.java
package ch.bailu.aat.views.preferences;

import android.content.Context;

import ch.bailu.aat.views.ImageButtonViewGroup;
import ch.bailu.aat_lib.logger.AppLog;
import ch.bailu.aat_lib.preferences.OnPreferencesChanged;
import ch.bailu.aat_lib.preferences.SolidIndexList;
import ch.bailu.aat_lib.preferences.StorageInterface;


public class SolidImageButton extends ImageButtonViewGroup implements OnPreferencesChanged {

    private final SolidIndexList solid;

    public SolidImageButton(Context context, SolidIndexList s) {
        super(context, s.getIconResource());

        solid = s;
        setOnClickListener(v -> {
            if (solid.length()<3) {
                solid.cycle();
            } else {
                new SolidIndexListDialog(context, solid);
            }

        });
    }


    @Override
    public void onAttachedToWindow() {
        super.onAttachedToWindow();
        setImageResource(solid.getIconResource());
        solid.register(this);
    }

    @Override
    public void onPreferencesChanged(StorageInterface s, String key) {
        if (solid.hasKey(key)) {
            setImageResource(solid.getIconResource());
            AppLog.i(getContext(), solid.getValueAsString());
        }
    }

    @Override
    public void onDetachedFromWindow() {
        super.onDetachedFromWindow();

        solid.unregister(this);
    }
}

File: aat-android/src/main/kotlin/ch/bailu/aat/views/preferences/SolidImageButton.kt
package ch.bailu.aat.views.preferences

import android.content.Context
import ch.bailu.aat.generated.Images
import ch.bailu.aat.views.image.ImageButtonViewGroup
import ch.bailu.aat.views.preferences.dialog.SolidIndexListDialog
import ch.bailu.aat_lib.logger.AppLog
import ch.bailu.aat_lib.preferences.OnPreferencesChanged
import ch.bailu.aat_lib.preferences.SolidIndexList
import ch.bailu.aat_lib.preferences.StorageInterface

class SolidImageButton(context: Context, private val solid: SolidIndexList) : ImageButtonViewGroup(
    context, Images.get(
        solid.getIconResource()
    )
), OnPreferencesChanged {
    init {
        setOnClickListener {
            if (solid.length() < 3) {
                solid.cycle()
            } else {
                SolidIndexListDialog(context, solid)
            }
        }
    }

    public override fun onAttachedToWindow() {
        super.onAttachedToWindow()
        setImageResource(Images.get(solid.getIconResource()))
        solid.register(this)
    }

    override fun onPreferencesChanged(storage: StorageInterface, key: String) {
        if (solid.hasKey(key)) {
            setImageResource(Images.get(solid.getIconResource()))
            AppLog.i(this, solid.getValueAsString())
        }
    }

    public override fun onDetachedFromWindow() {
        super.onDetachedFromWindow()
        solid.unregister(this)
    }
}
